relay-pattern/src/wildmatch.rs

```diff
@@ -0,0 +1,605 @@
+{
+    // Remainder of the haystack which still needs to be matched.
+    let mut h_current = haystack;
+    // Saved haystack position for backtracking.
+    let mut h_revert = haystack;
+    // Revert index for `tokens`. In case of backtracking we backtrack to this index.
+    //
+    // If `t_revert` is zero, it means there is no currently saved backtracking position.
+    let mut t_revert = 0;
+    // The next token position which needs to be evaluted.
+    let mut t_next = 0;
+
+    macro_rules! advance {
+        ($len:expr) => {{
+            h_current = &h_current[$len..];
+            true
+        }};
+    }
+
+    // Empty glob never matches.
+    if tokens.is_empty() {
+        return false;
+    }
+
+    while t_next != tokens.len() || !h_current.is_empty() {
+        let matched = if t_next == tokens.len() {
+            false
+        } else {
+            let token = &tokens[t_next];
+            t_next += 1;
+
+            match token {
+                Token::Literal(literal) => match M::is_prefix(h_current, literal) {
+                    Some(n) => advance!(n),
+                    // The literal does not match, but it may match after backtracking.
+                    // TODO: possible optimization: if the literal cannot possibly match anymore
+                    // because it is too long for the remaining haystack, we can immediately return
+                    // no match here.
+                    None => false,
+                },
+                Token::Any(n) => {
+                    advance!(match n_chars_to_bytes(*n, h_current) {
+                        Some(n) => n,
+                        // Not enough characters in the haystack remaining,
+                        // there cannot be any other possible match.
+                        None => return false,
+                    });
+                    true
+                }
+                Token::Wildcard => {
+                    // `ab*c*` matches `abcd`.
+                    if t_next == tokens.len() {
+                        return true;
+                    }
+
+                    t_revert = t_next;
+
+                    match skip_to_token::<_, M>(tokens, t_next, h_current) {
+                        Some((tokens, revert, remaining)) => {
+                            t_next += tokens;
+                            h_revert = revert;
+                            h_current = remaining;
+                        }
+                        None => return false,
+                    };
+
+                    true
+                }
+                Token::Class { negated, ranges } => match h_current.chars().next() {
+                    Some(next) => {
+                        M::ranges_match(next, *negated, ranges) && advance!(next.len_utf8())
+                    }
+                    None => false,
+                },
+                Token::Alternates(alternates) => {
+                    // TODO: should we make this iterative instead of recursive?
```

Comment:
I guess a malicious user could put a `[[[[[[[[[[[[[[[[a]]]]]]]]]]]]]]]]` in their config and make relay panic. But as long as that only crashes their own event processing it might still be fine.

Reply:
No, alternates themselves cannot be nested, `{{a}}` is a parsing error. What in theory is possible is `{..}{..}{..}{..}{..}{..}{..}{..}{..}{..}{..}{..}{..}{..}{..}..`.

In a follow-up I want to add track the complexity of a glob and limit it to an upper bound, which should address this case and the expensive backtracking combined with exponential search time of alternates.